src/flag_gems/ops/pad.py

```diff
@@ -0,0 +1,458 @@
+            code.writeline(f"{stride_args}, # stride for x")
+
+            # shape for inputs
+            for j in range(rank):
+                function_ns.create_name(f"out_strides{j}")
+            stride_args = ", ".join(f"out_strides{j}: int" for j in range(rank))
+            code.writeline(f"{stride_args}, # stride for out")
+
+            # shape for inputs
+            for j in range(rank):
+                function_ns.create_name(f"valid_dim{j}_start")
+            stride_args = ", ".join(f"valid_dim{j}_start: int" for j in range(rank))
+            code.writeline(f"{stride_args}, # valid dim start")
+
+            # shape for inputs
+            for j in range(rank):
+                function_ns.create_name(f"valid_dim{j}_end")
+            stride_args = ", ".join(f"valid_dim{j}_end: int" for j in range(rank))
+            code.writeline(f"{stride_args}, # valid dim end")
+
+            code.writeline("in_elem_cnt: tl.constexpr, ")
+            code.writeline("out_elem_cnt: tl.constexpr, ")
+            code.writeline("value, # padding value")
+            code.writeline("IS_CONSTANT: tl.constexpr, ")
+            code.writeline("IS_REFLECT: tl.constexpr, ")
+            code.writeline("IS_REPLICATE: tl.constexpr, ")
+            code.writeline("IS_CIRCULAR: tl.constexpr, ")
+            code.writeline("BLOCK_SIZE: tl.constexpr, ")
+
+    code.writeline("):")
+
+    with code.indent():
+        code.writeline("pid = tl.program_id(0)")
+        code.writeline("block_offset = pid * BLOCK_SIZE")
+        code.writeline("offset = block_offset + tl.arange(0, BLOCK_SIZE)")
+        code.newline()
+
+        code.writeline("remaining = offset ")
+        for i in range(rank):
+            code.writeline(f"idx = remaining // out_strides{i}")
+            code.writeline(f"dst_index_{i} = idx")
+            code.writeline(f"remaining = remaining - idx * out_strides{i}")
+            code.newline()
+
+        code.writeline("if_pad_false_mask = tl.zeros((BLOCK_SIZE, ), dtype=tl.int32)")
+        code.writeline("if_pad_true_mask = tl.full((BLOCK_SIZE, ), 1, dtype=tl.int32)")
+
+        for i in range(rank):
+            code.writeline(f"src_index_{i} = tl.zeros((BLOCK_SIZE, ), dtype=tl.int32)")
+
+        code.writeline(
+            "cond = (dst_index_0 >= valid_dim0_start and dst_index_0 < valid_dim0_end) "
+        )
+
+        for i in range(1, rank):
+            code.writeline(
+                f"cond &= (dst_index_{i} >= valid_dim{i}_start and dst_index_{i} < valid_dim{i}_end)"
+            )
+
+        code.writeline(
+            "if_pad = tl.where(cond, if_pad_false_mask, if_pad_true_mask).to(tl.int1)"
+        )
+
+        for i in range(rank):
+            code.writeline(f"src_index_{i} = dst_index_{i} - valid_dim{i}_start ")
+
+        for i in range(rank):
+            code.writeline(
+                f"src_index_{i} = tl.where(src_index_{i} < 0, 0, src_index_{i})"
+            )
```

Comment:
This would generate code like

```python
    if_pad_false_mask = tl.zeros((BLOCK_SIZE, ), dtype=tl.int32)
    if_pad_true_mask = tl.full((BLOCK_SIZE, ), 1, dtype=tl.int32)
    src_index_0 = tl.zeros((BLOCK_SIZE, ), dtype=tl.int32)
    src_index_1 = tl.zeros((BLOCK_SIZE, ), dtype=tl.int32)
    cond = (dst_index_0 >= valid_dim0_start and dst_index_0 < valid_dim0_end)
    cond &= (dst_index_1 >= valid_dim1_start and dst_index_1 < valid_dim1_end)
    if_pad = tl.where(cond, if_pad_false_mask, if_pad_true_mask).to(tl.int1)
    src_index_0 = dst_index_0 - valid_dim0_start
    src_index_1 = dst_index_1 - valid_dim1_start
    src_index_0 = tl.where(src_index_0 < 0, 0, src_index_0)
    src_index_1 = tl.where(src_index_1 < 0, 0, src_index_1)
```

if seems 

```python
    src_index_0 = tl.zeros((BLOCK_SIZE, ), dtype=tl.int32)
    src_index_1 = tl.zeros((BLOCK_SIZE, ), dtype=tl.int32)
```
is redundant.

Reply:
good catch, fixed.